interpn/src/multicubic/rectilinear.rs

```diff
@@ -63,16 +21,58 @@
 //! // Storage for output
 //! let mut out = [0.0; 2];
 //!
-//! // Do interpolation
+//! // Do interpolation, allocating for the output for convenience
 //! let linearize_extrapolation = false;
-//! rectilinear::interpn(grids, &z, linearize_extrapolation, &obs, &mut out).unwrap();
+//! rectilinear::interpn_alloc(grids, &z, linearize_extrapolation, &obs).unwrap();
 //! ```
 //!
 //! References
 //! * A. E. P. Veldman and K. Rinzema, “Playing with nonuniform grids”.
 //!   https://pure.rug.nl/ws/portalfiles/portal/3332271/1992JEngMathVeldman.pdf
 use num_traits::Float;
 
+/// Evaluate multicubic interpolation on a regular grid in up to 8 dimensions.
+/// Assumes C-style ordering of vals (z(x0, y0), z(x0, y1), ..., z(x0, yn), z(x1, y0), ...).
+///
+/// This is a convenience function; best performance will be achieved by using the exact right
+/// number for the MAXDIMS parameter, as this will slightly reduce compute and storage overhead,
+/// and the underlying method can be extended to more than this function's limit of 8 dimensions.
+/// The limit of 8 dimensions was chosen for no more specific reason than to reduce unit test times.
+///
+/// While this method initializes the interpolator struct on every call, the overhead of doing this
+/// is minimal even when using it to evaluate one observation point at a time.
+#[inline(always)]
```

Comment:
You probably don't want `inline(always)` here. In fact `inline(always)` is probably best avoided pretty much always.

Inlining in LLVM happens "leaves-first", which is backwards from how most of our intuition would expect it to happen. The LLVM inliner on its own would probably decide to inline the `::new()` call and the `.interp()` call _into this function_.

Then the `#[inline(always)]` will force it to inline this function into all call sites, and you'll end up with the contents of `MulticubicRectilinear::new()` and `MulticubicRectilinear::interp()` inlined all over the place _separately_. That's probably not desirable.

Reply:
Good point - each instance of inlining directives in here was originally checked in criterion benchmarks, but I'm finding lately that the benchmarks end up compiled somewhat differently from downstream usages. I'll check how it looks in end-to-end benchmarks via the python bindings with more reasonable inlining annotations.

Reply:
The perf gains from gratuitous inlining do appear to be an artifact of benchmarking. This snip (benchmarks running after switching from less-inlined version of both linear calcs back to unmodified) would indicate a significant perf regression for the specific case of evaluating one point at a time, which is an important case for differentiation:

![image](https://github.com/user-attachments/assets/91c713ac-7d1b-452b-add4-09f332709921)

However, end-to-end benchmarks through the python bindings do not show any significant degradation for the case of 1 observation point:

![image](https://github.com/user-attachments/assets/09128830-0b84-4c56-9a16-96a3e955f525)

Reply:
The rectilinear method doesn't show any significant change here, possibly because the core library bisection search is fairly low in the inlining tree, and is marked `inline(never)`

Reply:
No identifiable change in end-to-end benchmarks in python after removing all inlining annotations, even for the single-observation-point cases that see 2x slowdown in rust benchmarks. This should help embedded usages as well, since opt-level=s builds will behave more as expected

![image](https://github.com/user-attachments/assets/4461fae9-fcea-4743-8901-ecbcc07e1c8a)

Reply:
Inlining annotations removed everywhere

```
jlogan@jlogan-framework-13:~/git/interpn$ git grep inline
jlogan@jlogan-framework-13:~/git/interpn$ 
```

Reply:
Re: monomorphization bloat - definitely a thing that I think about. In this case, the only generics are on the data type and max dimensionality. The only types in common use that implement Float are f32 and f64 in core, plus f16 and f128 available in crates, and I don't anticipate seeing many uses of MAXDIMS other than the default of 8 since that's already an unreasonably large number of dimensions and doesn't affect perf noticeably.

Reply:
I think `#[inline]` is fine if you have leaf code you expect to be hot. It's really just `#[inline(always)]` especially on large non-leaf functions that ends up being problematic.

Reply:
Agreed - in this case, having just #[inline] doesn't produce any change in performance (even in the hypersensitive micro benchmarks), probably because the hot leaf functions are already small enough that they get inlined without adjusting the weights. So, might as well remove the manual annotations and just leave it to the compiler

Reply:
Thinking about this a little more - even in an opt-level=s build, you'd still want those functions to inline because they're tiny and the effect on perf would be really punishing if they didn't get inlined. So I'll sprinkle a few #[inline] back in there just in case

Reply:
Sprinkled some #[inline] on the chain of functions at the very bottom of the recursion for each interpolator type. No effect on benchmarks at all at opt-level=3, but might help keep something silly from happening in a different build someday